tests/virt_autotest/sriov_pci_passthrough.pm

```diff
@@ -0,0 +1,381 @@
+    script_run "lspci | grep Ethernet";
+  
+    #check host and guest to make sure they work well
+    if (check_host_and_guest($self, $guest)) {   #have not figured the parameters out???
+      $self->{test_results}->{$guest}->{"Error: host or guest $guest does not work well after vfs hot unplug from guest!"}->{status} = 'FAILED';
+    }
+
+  } #end each
+
+}
+
+
+#set up ssh, packages and iommu on host
+sub prepare_host(){
+
+  #generate ssh key on host
+  script_run "rm ~/.ssh/id_rsa*; ssh-keygen -t rsa -f ~/.ssh/id_rsa -P ''";
+
+  #install required packages on host 
+  zypper_call '-t in pciutils nmap';  #to run 'lspci' and 'nmap' ommand
+
+  #check IOMMU on XEN is enabled
+  if (get_var("XEN") || check_var("SYSTEM_ROLE", "xen") || check_var("HOST_HYPERVISOR", "xen")) {
+    assert_script_run "xl dmesg | grep IOMMU | grep -i enabled";
+  }
+}
+
+
+#get the BDF the PF device on host
+sub get_sriov_ethernet_devices () {
+
+  #get the BDF of the ethernet devices with SR-IOV
+  my $nic_devices = script_output "lspci | grep Ethernet | grep -v 'Virtual Function' | cut -d ' ' -f1" ;
+  my @nic_devices = split ("\n", $nic_devices);
+  my @sriov_devices;
+  foreach (@nic_devices){
+    if (! script_run "lspci -v -s $_ | grep -q 'SR-IOV'") {
+      push @sriov_devices, $_;
+    }
+  }
+  record_soft_failure "Error: there is no ethernet devices with SR-IOV in the host and test aborts!" if ! $#sriov_devices;
+  
+  return @sriov_devices;
+}
+
+#enable 8 virtual functions for the specified physical functions of the SR-IOV network device 
+sub enable_vf(){
+  my @pfs = @_;
+  
+  # get the network device drivers
+  my (@drivers, $pf, $driver);
+  foreach $pf (@pfs) {
+    $driver = script_output "lspci -v -s $pf | sed -n '/Kernel modules/p' | sed 's/.*Kernel modules: *//'";
+    my $is_known_driver = 0;
+    foreach (@drivers) {
+      next if ($_ ne "$driver");
+      $is_known_driver = 1;
+    }
+    push @drivers, $driver if ! $is_known_driver;
+  }
+  foreach (@drivers) {
+    assert_script_run("modprobe $driver", 60);
+  }
+ 
+  #set max_vfs and reload driver
+  #it seems the repeated enabling vf consused the system, or the some vf is not detached from a shutdown vm?
+  #so skip enabling in local test for debugging
+  record_info("Skip enabling VF");
+#  foreach $driver (@drivers) {
+#    assert_script_run("rmmod $driver", 30);   #4s last time
+#    assert_script_run("modprobe --first-time $driver max_vfs=8", 60);   #2s last time. but once fail with 30
+#  }
+
+  #bring up the SR-IOV device
+  foreach $pf (@pfs) {
+    my $nic = script_output "ls -l /sys/class/net |grep $pf | awk '{print \$9}'";
+    assert_script_run "echo \"BOOTPROTO='dhcp'\nSTARTMODE='manual'\" > /etc/sysconfig/network/ifcfg-$nic";
+    assert_script_run("ifup $nic", 60);   #15s last time
+  }
+  
+  my $vfs = script_output "lspci | grep Ethernet | grep \"Virtual Function\" | cut -d ' ' -f1";
+  my @vfs = split("\n", $vfs);
```

Comment:
Suggest use different variable name for its array and scalar variance.

Reply:
ok